src/function/scalar/string/string_split.cpp

```diff
@@ -0,0 +1,288 @@
+				idx_t i;
+				for (i = 1; i < delim_size; i++) {
+					if (input[offset + i] != delim[i])
+						break;
+				}
+				// delimiter found: skip start over delimiter
+				if (i == delim_size) {
+					start = offset + delim_size;
+					return offset;
+				}
+			}
+		}
+		return offset;
+	}
+
+protected:
+	const char *delim;
+	const size_t delim_size;
+};
+
+struct UnicodeIterator : virtual public Iterator {
+public:
+	UnicodeIterator(size_t input_size, const char *delim, const size_t delim_size)
+	    : Iterator(input_size), delim_size(delim_size) {
+		int cp_sz;
+		for (idx_t i = 0; i < delim_size; i += cp_sz) {
+			delim_cps.push_back(utf8proc_codepoint(delim, cp_sz));
+		}
+	}
+	idx_t Next(const char *input) override {
+		// special case: separate by empty delimiter
+		if (delim_size == 0) {
+			offset = utf8proc_next_grapheme(input, size, offset);
+			start = offset;
+			return offset;
+		}
+		int cp_sz;
+		for (offset = start; HasNext(); offset = utf8proc_next_grapheme(input, size, offset)) {
+			// potential delimiter match
+			if (utf8proc_codepoint(&input[offset], cp_sz) == delim_cps[0] && offset + delim_size <= size) {
+				idx_t delim_offset = cp_sz;
+				for (idx_t i = 1; i < delim_cps.size(); i++) {
+					if (utf8proc_codepoint(&input[offset + delim_offset], cp_sz) != delim_cps[i])
+						break;
+					delim_offset += cp_sz;
+				}
+				// delimiter found: skip start over delimiter
+				if (delim_offset == delim_size) {
+					start = offset + delim_size;
+					return offset;
+				}
+			}
+		}
+		return offset;
+	}
+
+protected:
+	vector<utf8proc_int32_t> delim_cps;
+	const size_t delim_size;
+};
+
+struct RegexIterator : virtual public Iterator {
+public:
+	RegexIterator(size_t input_size, unique_ptr<RE2> re, const bool ascii_only)
+	    : Iterator(input_size), re(move(re)), ascii_only(ascii_only) {
+	}
+	idx_t Next(const char *input) override {
+		duckdb_re2::StringPiece input_sp(input, size);
+		duckdb_re2::StringPiece match;
+		if (re->Match(input_sp, start, size, RE2::UNANCHORED, &match, 1)) {
+			offset = match.data() - input;
+			// special case: 0 length match
+			if (match.size() == 0) {
+				if (ascii_only)
+					offset++;
+				else
+					offset = utf8proc_next_grapheme(input, size, offset);
+				start = offset;
+			} else {
+				start = offset + match.size();
+			}
+		} else {
+			offset = size;
+		}
+		return offset;
+	}
+
+protected:
+	unique_ptr<RE2> re;
+	const bool ascii_only;
+};
+
+void string_split(const char *input, Iterator &iter, ChunkCollection &result) {
+	auto append_chunk = make_unique<DataChunk>();
+	vector<LogicalType> types = {LogicalType::VARCHAR};
+	append_chunk->Initialize(types);
+
+	// special case: empty string
+	if (iter.size == 0) {
+		FlatVector::GetData<string_t>(append_chunk->data[0])[append_chunk->size()] =
+		    StringVector::AddString(append_chunk->data[0], &input[0], 0);
+		append_chunk->SetCardinality(append_chunk->size() + 1);
+	}
+
+	while (iter.HasNext()) {
```

Comment:
This condition needs not to be checked if `iter.size == 0`

Reply:
Changed so that if `iter.size == 0` it will add `append_chunk` and `return`